tests/test_build.py

```diff
@@ -35,3 +41,94 @@ def test_build_project_metadata_resolved_correct_build_dependencies(
         "setuptools",
         "wheel",
     ]
+
+
+def test_build_project_metadata_static(tmp_path):
+    """Test static parsing branch of build_project_metadata"""
+    src_pkg_path = pathlib.Path(PACKAGES_PATH) / "small_fake_with_pyproject"
+    shutil.copytree(src_pkg_path, tmp_path, dirs_exist_ok=True)
+    src_file = tmp_path / "pyproject.toml"
+    metadata = build_project_metadata(
+        src_file, (), attempt_static_parse=True, isolated=True, quiet=False
+    )
+    assert isinstance(metadata, StaticProjectMetadata)
+    requirements = [(r.name, r.extras, str(r.markers)) for r in metadata.requirements]
+    requirements.sort(key=lambda x: x[0])
+    assert requirements == [
+        ("fake_direct_extra_runtime_dep", {"with_its_own_extra"}, 'extra == "x"'),
+        ("fake_direct_runtime_dep", set(), "None"),
+    ]
+    assert metadata.extras == ("x",)
+
+
+def test_build_project_metadata_raises_error(tmp_path):
+    src_pkg_path = pathlib.Path(PACKAGES_PATH) / "small_fake_with_build_deps"
+    shutil.copytree(src_pkg_path, tmp_path, dirs_exist_ok=True)
+    src_file = tmp_path / "setup.py"
+    with pytest.raises(
+        ValueError, match="Cannot execute the PEP 517 optional.* hooks statically"
+    ):
+        build_project_metadata(
+            src_file,
+            ("editable",),
+            attempt_static_parse=True,
+            isolated=True,
+            quiet=False,
+        )
+
+
+def test_static_parse(tmp_path):
+    src_file = tmp_path / "pyproject.toml"
+
+    valid = """
+[project]
+name = "foo"
+version = "0.1.0"
+dependencies = ["bar>=1"]
+[project.optional-dependencies]
+baz = ["qux[extra]"]
+"""
+    src_file.write_text(valid)
+    metadata = maybe_statically_parse_project_metadata(src_file)
+    assert isinstance(metadata, StaticProjectMetadata)
+    assert [str(r.req) for r in metadata.requirements] == ["bar>=1", "qux[extra]"]
+    assert metadata.extras == ("baz",)
+
+    invalid_toml = """this is not valid toml"""
+    src_file.write_text(invalid_toml)
+    assert maybe_statically_parse_project_metadata(src_file) is None
+
+    no_pep621 = """
+[build-system]
+requires = ["setuptools"]
+"""
+    src_file.write_text(no_pep621)
+    assert maybe_statically_parse_project_metadata(src_file) is None
```

Comment:
These are quite a lot `assert` statements. How about creating a tuple of src files and running a for loop on them at the end of the test?

Reply:
I think that might make it harder to figure out which one is failing (especially since the file name needs to be pyproject.toml). I split up the test a little bit though!